File: cpp/tree/src/tree.cpp
#include <iostream>
#include <string>
#include <vector>
#include <filesystem>

namespace fs = std::filesystem;
using namespace std;

void print_backing(string_view subbed, bool dir, int len) {
  cout << '|';
  for(int i = 0; i < len; i++) {
    cout << "  |";
  }
  cout << "--";
  if(dir) {
    cout << "/";
  }
  cout << subbed << endl;
}

void print_dir(string dir, int len) {
  vector<string> execs;
  vector<string> hdirs;
  for(const auto &f : fs::directory_iterator(dir)) {
    string path = f.path();
    string subbed = fs::path(path).filename();
    if(fs::is_directory(path)) {
      if(subbed[0] != '.') {
         print_backing(subbed, true, len);
	 print_dir(path, len + 1);
      }else{
         hdirs.push_back(subbed);
      }
    } else {
      execs.push_back(subbed);
    }
  }
  for(const auto &hd : hdirs) {
    print_backing(hd + " (content hidden)", true, len);
  }
  for(const auto &e : execs) {
    print_backing(e, false, len);
  }
}

int main(int argc, char *argv[]) {
  string file;
  if (argc > 1) {
    file = argv[1];
    if(!fs::is_directory(file)) {
      cout << "Failed to find `" << file << "` as a valid directory." << endl;
      return 1;
    }
  } else {
    file = fs::current_path();
  }
  cout << file << endl;
  print_dir(file, 0);
  return 0;
}

File: python/tree/tree.py
#!/bin/python
import os;
import sys;

def write_file(where, is_dir, length):
  out = "|"
  for _ in range(0, length):
    out += "  |"
  out += "--"
  if is_dir: out += "/"
  out += where
  print(out)

def walk_dir(directory, length):
  hdirs = list()
  execs = list()
  for f in os.listdir(directory):
    if os.path.isdir(os.path.join(directory, f)):
      if f[0] == '.':
        hdirs.append(f)
      else:
        write_file(f, True, length)
        walk_dir(os.path.join(directory, f), length + 1)       
    else:
      execs.append(f)
  [write_file(x + " (content hidden)", True, length) for x in hdirs]
  [write_file(x, False, length) for x in execs]

where = os.getcwd() if len(sys.argv) == 1 else sys.argv[1]
if os.path.isdir(where):
  print(where)
  walk_dir(where, 0)
else:
  print(f"{where} is not a valid directory.")
